django/db/models/fields/related_descriptors.py

```diff
@@ -225,21 +225,31 @@ def __get__(self, instance, cls=None):
         try:
             rel_obj = self.field.get_cached_value(instance)
         except KeyError:
+            rel_obj = None
             has_value = None not in self.field.get_local_related_value(instance)
-            ancestor_link = (
-                instance._meta.get_ancestor_link(self.field.model)
-                if has_value
-                else None
-            )
-            if ancestor_link and ancestor_link.is_cached(instance):
-                # An ancestor link will exist if this field is defined on a
-                # multi-table inheritance parent of the instance's class.
-                ancestor = ancestor_link.get_cached_value(instance)
-                # The value might be cached on an ancestor if the instance
-                # originated from walking down the inheritance chain.
-                rel_obj = self.field.get_cached_value(ancestor, default=None)
-            else:
-                rel_obj = None
+            current_instance = instance
+            keep_going = has_value
+            while keep_going:
+                ancestor_link = current_instance._meta.get_ancestor_link(
+                    self.field.model
+                )
+                if ancestor_link and ancestor_link.is_cached(current_instance):
+                    # An ancestor link will exist if this field is defined on a
+                    # multi-table inheritance parent of the instance's class.
+                    ancestor = ancestor_link.get_cached_value(current_instance)
```

Comment:
Looking at the implementation of `get_cached_value` we could likely avoid making a second call to `is_cached` here.

```suggestion
                if not (
                    # An ancestor link will exist if this field is defined on a
                    # multi-table inheritance parent of the instance's class.
                    ancestor_link
                    and (
                        ancestor := ancestor_link.get_cached_value(current_instance, None)
                    ) is not None
                 ):
```

Reply:
Absolutely true, I have implemented this for all three cases where it happened.